File: ec_draw/diagram.py
"""High-level Diagram builder.

The Diagram class is the main entry point for creating Excalidraw diagrams.
It tracks elements, applies themes and layouts, and renders to .excalidraw files.

Usage:
    from ec_draw import Diagram

    d = Diagram(theme="sketchy")
    d.add_box("API Gateway", row=0, col=0, span=3)
    d.add_box("Auth Service", row=1, col=0)
    d.add_box("Database", row=1, col=1)
    d.add_arrow("API Gateway", "Auth Service")
    d.add_arrow("Auth Service", "Database")
    d.save("architecture.excalidraw")
"""

import json
import os
from typing import Optional

from .elements import (
    rect,
    ellipse,
    diamond,
    arrow,
    line,
    text_label,
    frame,
    _make_id,
)
from .themes import Theme, get_theme, list_themes
from .layout import GridLayout


class Diagram:
    """Build an Excalidraw diagram with a consistent theme and layout."""

    def __init__(
        self,
        theme: str = "sketchy",
        *,
        cols: int = 3,
        cell_w: int = 160,
        cell_h: int = 80,
        gap_x: int = 40,
        gap_y: int = 50,
    ):
        """Create a new diagram.

        Args:
            theme: theme name ("sketchy", "professional", "dark", "colorful")
            cols: default grid columns
            cell_w, cell_h: default cell size for grid layout
            gap_x, gap_y: spacing between cells
        """
        self.theme: Theme = get_theme(theme) if isinstance(theme, str) else theme
        self.elements: list[dict] = []
        self._extra_elements: list[dict] = []  # text labels, etc.
        self._named: dict[str, dict] = {}  # name → element lookup
        self._grid = GridLayout(
            cols=cols,
            cell_w=cell_w,
            cell_h=cell_h,
            gap_x=gap_x,
            gap_y=gap_y,
        )
        self._color_idx = 0

    # ── Adding elements ───────────────────────────────────────────

    def add_box(
        self,
        name: str,
        *,
        row: Optional[int] = None,
        col: Optional[int] = None,
        span: int = 1,
        width: Optional[float] = None,
        height: Optional[float] = None,
        shape: str = "rect",
    ) -> dict:
        """Add a labeled box (rectangle) to the diagram.

        Args:
            name: label text and lookup key
            row, col: grid position (auto-increments if omitted)
            span: how many columns this box spans
            width, height: override default cell size
            shape: "rect", "ellipse", or "diamond"

        Returns:
            The element dict for the shape
        """
        if row is None and col is None:
            # auto-position
            idx = len(self._named)
            row = idx // self._grid.cols
            col = idx % self._grid.cols
        elif row is None:
            row = 0
        elif col is None:
            col = 0

        x, y = self._grid.position(row * self._grid.cols + col)
        w = width or (self._grid.cell_w + (span - 1) * (self._grid.cell_w + self._grid.gap_x))
        h = height or self._grid.cell_h

        builders = {
            "rect": rect,
            "ellipse": ellipse,
            "diamond": diamond,
        }
        build = builders.get(shape, rect)

        elem = build(
            x, y, w, h,
            text=name,
            theme=self.theme,
            color_index=self._color_idx,
        )
        self._color_idx += 1
        self.elements.append(elem)
        self._named[name] = elem

        # Add bound text element
        if name:
            text_elem = text_label(
                x + w / 2 - len(name) * self.theme.font_size * 0.3,
                y + h / 2 - self.theme.font_size * 0.6,
                name,
                theme=self.theme,
                container_id=elem["id"],
            )
            self.elements.append(text_elem)

        return elem

    def add_arrow(
        self,
        from_name: str,
        to_name: str,
        *,
        label: Optional[str] = None,
        start_arrowhead: Optional[str] = None,
        end_arrowhead: Optional[str] = "arrow",
    ) -> dict:
        """Add an arrow connecting two named boxes.

        Args:
            from_name: name of the source box
            to_name: name of the target box
            label: optional text on the arrow
            start_arrowhead: arrowhead at start (None, "arrow", "bar", "dot", "triangle")
            end_arrowhead: arrowhead at end

        Returns:
            The arrow element dict
        """
        from_elem = self._named.get(from_name)
        to_elem = self._named.get(to_name)

        if not from_elem:
            raise KeyError(f"No element named '{from_name}'. Available: {list(self._named.keys())}")
        if not to_elem:
            raise KeyError(f"No element named '{to_name}'. Available: {list(self._named.keys())}")

        # Draw from center-right of from_elem to center-left of to_elem
        fx = from_elem["x"] + from_elem["width"]
        fy = from_elem["y"] + from_elem["height"] / 2
        tx = to_elem["x"]
        ty = to_elem["y"] + to_elem["height"] / 2

        elem = arrow(
            fx, fy, tx, ty,
            label=label,
            theme=self.theme,
            start_arrowhead=start_arrowhead,
            end_arrowhead=end_arrowhead,
        )
        self.elements.append(elem)
        return elem

    def add_text(
        self,
        content: str,
        x: float,
        y: float,
        *,
        font_size: Optional[int] = None,
    ) -> dict:
        """Add a standalone text element.

        Args:
            content: the text to display
            x, y: position
            font_size: override theme default

        Returns:
            The text element dict
        """
        fs = font_size or self.theme.font_size
        elem = text_label(x, y, content, theme=self.theme, fontSize=fs)
        self.elements.append(elem)
        return elem

    def add_line(
        self,
        x1: float,
        y1: float,
        x2: float,
        y2: float,
    ) -> dict:
        """Add a plain line (no arrowheads).

        Args:
            x1, y1: start point
            x2, y2: end point

        Returns:
            The line element dict
        """
        elem = line(x1, y1, x2, y2, theme=self.theme)
        self.elements.append(elem)
        return elem

    def add_frame(
        self,
        name: str,
        x: float,
        y: float,
        width: float,
        height: float,
    ) -> dict:
        """Add a frame (grouping container) around elements.

        Args:
            name: frame label
            x, y: position
            width, height: dimensions

        Returns:
            The frame element dict
        """
        elem = frame(x, y, width, height, name, theme=self.theme)
        self.elements.append(elem)
        return elem

    def add_element(self, element: dict) -> dict:
        """Add a raw element dict directly.

        Useful for custom elements not covered by the convenience methods.

        Returns:
            The element dict (same as input)
        """
        self.elements.append(element)
        return element

    # ── Rendering ──────────────────────────────────────────────────

    def to_dict(self) -> dict:
        """Return the complete .excalidraw document as a dict."""
        return {
            "type": "excalidraw",
            "version": 2,
            "source": "https://excalidraw.com",
            "elements": self.elements,
            "appState": {
                "gridSize": None,
                "viewBackgroundColor": self.theme.colors.background,
                "currentItemFontFamily": self.theme.font_family,
                "currentItemFontSize": self.theme.font_size,
                "theme": "dark" if self.theme.name == "dark" else "light",
            },
            "files": {},
        }

    def to_json(self, indent: int = 2) -> str:
        """Return the .excalidraw document as a JSON string."""
        return json.dumps(self.to_dict(), ensure_ascii=False, indent=indent)

    def save(self, path: str) -> str:
        """Save the diagram as a .excalidraw file.

        Args:
            path: output file path (should end with .excalidraw)

        Returns:
            The absolute path to the saved file
        """
        if not path.endswith(".excalidraw"):
            path += ".excalidraw"

        doc = self.to_dict()
        with open(path, "w", encoding="utf-8") as f:
            json.dump(doc, f, ensure_ascii=False, indent=2)

        abs_path = os.path.abspath(path)
        return abs_path

    # ── Utilities ──────────────────────────────────────────────────

    def __len__(self) -> int:
        return len(self.elements)

    def __repr__(self) -> str:
        return f"Diagram(theme={self.theme.name!r}, elements={len(self.elements)})"

File: src/diagram.ts
/**
 * Diagram builder — programmatic API for creating Excalidraw diagrams.
 *
 * Use this for freeform layouts (architecture diagrams, whiteboard sketches)
 * where you need full control over element placement.
 *
 * For structured diagrams (flowcharts, sequences, ER), prefer renderDiagram()
 * with JSON descriptors and automatic layout.
 */

import { writeFileSync } from "fs";
import { resolve } from "path";
import type { ExcalidrawElement, ExcalidrawDocument, ShapeType } from "./types.js";
import { getTheme, type ThemeConfig } from "./themes.js";
import { normalizeElement, createElement, createTextElement, buildAppState, textWidth } from "./normalize.js";
import { ICONS, loadLibraryIcon } from "./library.js";
import type { IconDef } from "./library.js";
import { gridPosition } from "./layout/grid.js";
import { Card, Section, Callout, type Component } from "./components.js";

export interface BoxOptions {
  row?: number;
  col?: number;
  span?: number;
  width?: number;
  height?: number;
  shape?: ShapeType;
}

export class Diagram {
  theme: ThemeConfig;
  elements: ExcalidrawElement[];
  private _named: Map<string, ExcalidrawElement>;
  private _colorIdx: number;
  private _cols: number;
  private _cellW: number;
  private _cellH: number;
  private _gapX: number;
  private _gapY: number;

  constructor(
    theme: string | ThemeConfig = "sketchy",
    opts: {
      cols?: number;
      cellW?: number;
      cellH?: number;
      gapX?: number;
      gapY?: number;
    } = {}
  ) {
    this.theme = typeof theme === "string" ? getTheme(theme) : theme;
    this.elements = [];
    this._named = new Map();
    this._colorIdx = 0;
    this._cols = opts.cols ?? 3;
    this._cellW = opts.cellW ?? 160;
    this._cellH = opts.cellH ?? 80;
    this._gapX = opts.gapX ?? 40;
    this._gapY = opts.gapY ?? 50;
  }

  /** Resolve role-based colors. Falls back to shape palette cycling. */
  private _resolveRoleColor(role?: string): [string, string] {
    if (role && this.theme.roles && this.theme.roles[role]) {
      return this.theme.roles[role];
    }
    const [stroke, bg] = this.theme.shapes[this._colorIdx % this.theme.shapes.length];
    this._colorIdx++;
    return [stroke, bg];
  }

  private _gridPos(index: number): { x: number; y: number } {
    return gridPosition(index, this._cols, this._cellW, this._cellH, this._gapX, this._gapY);
  }

  /** Add a labeled box to the diagram. */
  addBox(
    name: string,
    opts: BoxOptions = {}
  ): ExcalidrawElement {
    let idx: number;
    if (opts.row !== undefined && opts.col !== undefined) {
      idx = opts.row * this._cols + opts.col;
    } else {
      idx = this._named.size;
    }

    const { x, y } = this._gridPos(idx);
    const span = opts.span ?? 1;
    const w = opts.width ?? (this._cellW + (span - 1) * (this._cellW + this._gapX));
    const h = opts.height ?? this._cellH;
    const shape = opts.shape ?? "rectangle";

    const [stroke, bg] = this.theme.shapes[this._colorIdx % this.theme.shapes.length];
    this._colorIdx++;

    const el = normalizeElement(
      createElement(shape, {
        x, y, width: w, height: h,
        strokeColor: stroke, backgroundColor: bg,
        fillStyle: this.theme.fillStyle,
        strokeWidth: this.theme.strokeWidth,
        strokeStyle: this.theme.strokeStyle,
        roughness: this.theme.roughness,
        roundness: this.theme.roundness,
      }),
      this.theme,
      this._colorIdx - 1
    );

    this.elements.push(el);
    this._named.set(name, el);

    // Bound text label
    if (name) {
      const textEl = normalizeElement(
        createTextElement(
          name,
          x + w / 2 - textWidth(name, this.theme.fontSize) / 2,
          y + h / 2 - this.theme.fontSize * 0.6,
          this.theme.fontSize,
          this.theme.fontFamily,
          el.id
        ),
        this.theme,
        0
      ) as any;
      textEl.strokeColor = this.theme.text;
      textEl.containerId = el.id;
      textEl.originalText = name;

      el.boundElements = [{ id: textEl.id, type: "text" }];
      this.elements.push(textEl);
    }

    return el;
  }

  /** Connect two named boxes with an arrow. */
  addArrow(
    fromName: string,
    toName: string,
    opts: { label?: string } = {}
  ): ExcalidrawElement {
    const from = this._named.get(fromName);
    const to = this._named.get(toName);
    if (!from) throw new Error(`No element named "${fromName}"`);
    if (!to) throw new Error(`No element named "${toName}"`);

    const fx = from.x + from.width;
    const fy = from.y + from.height / 2;
    const tx = to.x;
    const ty = to.y + to.height / 2;
    const dx = tx - fx;
    const dy = ty - fy;

    const el = normalizeElement(
      createElement("arrow", {
        x: fx, y: fy, width: dx, height: dy,
        strokeColor: this.theme.arrow,
        roundness: null,
      }),
      this.theme,
      0
    ) as any;
    el.points = [[0, 0], [dx, dy]];
    el.startArrowhead = null;
    el.endArrowhead = "arrow";
    el.startBinding = null;
    el.endBinding = null;

    if (opts.label) {
      const textEl = normalizeElement(
        createTextElement(
          opts.label,
          fx + dx / 2 - textWidth(opts.label, 14) / 2,
          fy + dy / 2 - this.theme.fontSize * 1.2,
          14,
          this.theme.fontFamily,
          el.id
        ),
        this.theme,
        0
      ) as any;
      textEl.strokeColor = this.theme.text;
      textEl.containerId = el.id;
      textEl.originalText = opts.label;

      el.boundElements = [{ id: textEl.id, type: "text" }];
      this.elements.push(textEl);
    }

    this.elements.push(el);
    return el;
  }

  /** Add a standalone text element. */
  addText(content: string, x: number, y: number, fontSize?: number): ExcalidrawElement {
    const fs = fontSize ?? this.theme.fontSize;
    const el = normalizeElement(
      createTextElement(content, x, y, fs, this.theme.fontFamily),
      this.theme,
      0
    ) as any;
    el.strokeColor = this.theme.text;
    el.originalText = content;
    this.elements.push(el);
    return el;
  }

  /** Place an icon from the library at a position. */
  addIcon(
    iconName: string,
    x: number,
    y: number,
    colorIndex: number = 0
  ): void {
    const icon = ICONS[iconName];
    if (!icon) {
      throw new Error(`Unknown icon "${iconName}". Available: ${Object.keys(ICONS).join(", ")}`);
    }
    const [stroke, bg] = this.theme.shapes[colorIndex % this.theme.shapes.length];
    const elements = icon.render(x, y, stroke, bg);
    for (const el of elements) {
      this.elements.push(normalizeElement(el, this.theme, colorIndex));
    }
  }

  /** Place an icon from an external .excalidrawlib file at a position. */
  addLibraryIcon(
    libraryName: string,
    iconName: string,
    x: number,
    y: number,
    colorIndex: number = 0
  ): void {
    const [stroke, bg] = this.theme.shapes[colorIndex % this.theme.shapes.length];
    const elements = loadLibraryIcon(libraryName, iconName, x, y, { stroke, bg });
    for (const el of elements) {
      this.elements.push(normalizeElement(el, this.theme, colorIndex));
    }
  }

  /** Add a sequence diagram actor using stick figure from library. */
  addActor(
    name: string,
    x: number,
    y: number,
    libraryName: string = "stick-figures",
    iconName: string = "Stick man"
  ): void {
    this.addLibraryIcon(libraryName, iconName, x, y, 0);
    this.addText(name,
      x + 30 - textWidth(name, this.theme.fontSize) / 2,
      y + 80, this.theme.fontSize);
  }

  // ── Semantic Components ──────────────────────────────────────

  /** Add a semantic card — title + subtitle + optional icon. */
  addCard(
    id: string,
    opts: {
      title: string;
      subtitle?: string;
      icon?: string;
      role?: string;
      row?: number;
      col?: number;
      span?: number;
      width?: number;
      height?: number;
    }
  ): Component {
    const idx = opts.row !== undefined && opts.col !== undefined
      ? opts.row * this._cols + opts.col
      : this._named.size;
    const { x, y } = this._gridPos(idx);
    const span = opts.span ?? 1;
    const w = opts.width ?? (this._cellW + (span - 1) * (this._cellW + this._gapX));
    const h = opts.height ?? this._cellH;

    const [stroke, bg] = this._resolveRoleColor(opts.role);

    let iconDef: IconDef | undefined;
    if (opts.icon) {
      iconDef = ICONS[opts.icon];
      if (!iconDef) {
        throw new Error(`Unknown icon "${opts.icon}". Available: ${Object.keys(ICONS).join(", ")}`);
      }
    }

    const component = Card(x, y, w, h, opts.title, stroke, bg, this.theme, {
      subtitle: opts.subtitle,
      icon: iconDef,
      role: opts.role,
    });

    this._named.set(id, component.primary);
    this.elements.push(...component.elements);
    return component;
  }

  /** Add a visual section — container frame with heading. */
  addSection(
    id: string,
    opts: {
      heading: string;
      subtitle?: string;
      role?: string;
      row?: number;
      col?: number;
      span?: number;
      width?: number;
      height?: number;
    }
  ): Component {
    const idx = opts.row !== undefined && opts.col !== undefined
      ? opts.row * this._cols + opts.col
      : this._named.size;
    const { x, y } = this._gridPos(idx);
    const span = opts.span ?? 1;
    const w = opts.width ?? (this._cellW + (span - 1) * (this._cellW + this._gapX));
    const h = opts.height ?? this._cellH;
    if (h < 60) throw new Error(`Section "${id}" height must be ≥ 60px (got ${h}).`);

    const [stroke, bg] = this._resolveRoleColor(opts.role);

    const component = Section(x, y, w, h, opts.heading, stroke, bg, this.theme, {
      subtitle: opts.subtitle,
      role: opts.role,
    });

    this._named.set(id, component.primary);
    this.elements.push(...component.elements);
    return component;
  }

  /** Add a full-width callout — emphasis bar with icon + text. */
  addCallout(
    id: string,
    opts: {
      text: string;
      icon?: string;
      role?: string;
      row?: number;
      col?: number;
      span?: number;
    }
  ): Component {
    const idx = opts.row !== undefined && opts.col !== undefined
      ? opts.row * this._cols + opts.col
      : this._named.size;
    const { x, y } = this._gridPos(idx);
    const span = opts.span ?? this._cols;
    const w = (this._cellW + this._gapX) * span - this._gapX;

    const [stroke, bg] = this._resolveRoleColor(opts.role);

    let iconDef: IconDef | undefined;
    if (opts.icon) {
      iconDef = ICONS[opts.icon];
      if (!iconDef) {
        throw new Error(`Unknown icon "${opts.icon}". Available: ${Object.keys(ICONS).join(", ")}`);
      }
    }

    const component = Callout(x, y, w, opts.text, stroke, bg, this.theme, {
      icon: iconDef,
      role: opts.role,
    });

    this._named.set(id, component.primary);
    this.elements.push(...component.elements);
    return component;
  }

  /** Add a labeled transition arrow. Label is required — use addArrow() for unlabeled arrows. */
  addTransition(
    fromId: string,
    toId: string,
    opts: { label: string }
  ): ExcalidrawElement {
    if (!opts.label || opts.label.trim() === "") {
      throw new Error(
        `Transition from "${fromId}" to "${toId}" requires a narrative label. Use addArrow() for unlabeled arrows.`
      );
    }
    return this.addArrow(fromId, toId, opts);
  }

  // ── Serialization ────────────────────────────────────────────

  /** Convert to a plain object (for JSON serialization). */
  toDocument(): ExcalidrawDocument {
    return {
      type: "excalidraw",
      version: 2,
      source: "https://excalidraw.com",
      elements: this.elements,
      appState: buildAppState(this.theme),
      files: {},
    };
  }

  /** Serialize to JSON string. */
  toJSON(): string {
    return JSON.stringify(this.toDocument(), null, 2);
  }

  /** Save to a .excalidraw file. */
  save(filepath: string): string {
    const path = filepath.endsWith(".excalidraw") ? filepath : filepath + ".excalidraw";
    writeFileSync(path, this.toJSON(), "utf-8");
    return resolve(path);
  }
}
